How many late days are you using for this assignment? 

0

In [ ]:
# Imports and pip installations (if needed)
import pandas as pd



# Part 1: Load the dataset

In [43]:
# Load the given datasets
dataset1 = pd.read_csv('/Users/<redacted>/Downloads/chronic_kidney_disease_categorical.csv')
dataset2 = pd.read_csv('/Users/<redacted>/Downloads/chronic_kidney_disease_numerical.csv')

# Print the data
print(dataset1)
print(dataset2)



     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

Numerical -> age, blood pressure, blood glucose random, blood urea, serum creatinine, sodium, potassium, hemoglobin, packed cell volume, white blood cell count, red blood cell count
This data set is numerical, every category is represented by numbers like the it's the value of blood pressure and the white blood cell count which are all represented by numbers. 
Numerical columns: blood pressure (bp), blood glucose random (bgr), sodium(sod)
 

Categorical -> albumin, sugar, red blood cells, pus cell, pus cell clumps, bacteria, hypertension, diabetes mellitus, coronary artery disease, appetite, pedal edema, anemia, Target 
There are categorical and numerical features in this dataset, but mainly categorical. Thre are only two categories that are numerical, albumin and sugar. In this dataset, the categorical features have only two options like yes or no, good or poor, present and notpresent.
Numerical columns: albumin (al), sugar(su)
Categorical columns: bacteria(ba), hypertension(htn), coronary artery disease(cad)


# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [44]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates1 = dataset1.duplicated(subset=["unique_id"])
if (duplicates1.sum() > 0):
    print(duplicates1.sum())

# Drop these duplicate rows
duplicates1Cleaned = dataset1.drop_duplicates()
#checking if duplicates rows were actually removed
duplicates1Check = duplicates1Cleaned.duplicated()
print(duplicates1Check.sum())



# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates2 = dataset2.duplicated(subset=["unique_id"])
if (duplicates2.sum() > 0):
    print(duplicates2.sum())
duplicates2Cleaned = dataset2.drop_duplicates()
#checking if duplicates rows were actually removed
duplicates2Check = duplicates2Cleaned.duplicated()
print(duplicates2Check.sum())



30
0
30
0


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Not all unique ids are present in both sets because the two datsets have different dimensions (shown below) and these dataset are already cleaned of outliers so it's impossible for all unique ids to be present in both data sets. The rows that are missing from one of the datsets just have a lot of NAN(null) values where the data from that datset would be.

In [57]:
print(duplicates1Cleaned.shape)
print(duplicates2Cleaned.shape)
# Merge the two given numerical and categorical datasets based on their unique_ID.
mergedDataset = duplicates1Cleaned.merge(duplicates2Cleaned, on = "unique_id", how = "outer")


#Print the combined dataset
print(mergedDataset)


(400, 14)
(390, 12)
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1       103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2       104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3       105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4       107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
397     996412  NaN  NaN     NaN       NaN  notpresent  notpresent   no   no   
398     997629  0.0  0.0  normal    normal         NaN         NaN   no   no   
399     997876  0.0  0.0     NaN    normal  notpresent  notpresent   no   no   

     cad  ...    bp

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [46]:
# Calculate the percentage of rows that contain atleast one missing value
totalRows = mergedDataset.shape[0]
missingRows = mergedDataset.isnull().any(axis=1).sum()
missingPercentage = (missingRows/totalRows)  * 100
# Print %
print(missingPercentage)
# Drop these rows from the dataset
datasetCleaned = mergedDataset.dropna()
# Print the Dataset
print(datasetCleaned)

61.75000000000001
     unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0       102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
6       109053  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
7       114717  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
8       118191  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
9       118596  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
..         ...  ...  ...     ...       ...         ...         ...  ...  ...   
390     980291  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  yes   
391     991031  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
394     994377  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
395     995177  4.0  3.0  normal  abnormal     present     present  yes  yes   
396     996076  0.0  0.0  normal    normal  notpresent  notpresent   no   no   

     cad  ...    bp  

## Part 3.4: Sort the dataset according to the Labels

In [47]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sortedDataset = datasetCleaned.sort_values(by = "Target", ascending=True)
sortedDataset = sortedDataset.reset_index(drop=True)
# Print the dataset
print(sortedDataset)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   
1       337834  4.0  0.0    normal  abnormal     present  notpresent  yes   
2       343710  2.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       349892  1.0  3.0  abnormal  abnormal  notpresent  notpresent  yes   
4       353025  3.0  4.0    normal    normal  notpresent  notpresent  yes   
..         ...  ...  ...       ...       ...         ...         ...  ...   
148     486397  0.0  0.0    normal    normal  notpresent  notpresent   no   
149     203466  0.0  0.0    normal    normal  notpresent  notpresent   no   
150     482105  0.0  0.0    normal    normal  notpresent  notpresent   no   
151     228849  0.0  0.0    normal    normal  notpresent  notpresent   no   
152     996076  0.0  0.0    normal    normal  notpresent  notpresent   no   

      dm  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wb

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [48]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encodedData = pd.get_dummies(sortedDataset, columns=categorical_columns, dtype=int)
#print(encodedData.columns.tolist())

# Print the dataset
print(encodedData)


     unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
0       102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0  7.6  ...   
1       337834  4.0  0.0  48.0  70.0  117.0   56.0   3.8  111.0  2.5  ...   
2       343710  2.0  0.0  61.0  80.0  173.0  148.0   3.9  135.0  5.2  ...   
3       349892  1.0  3.0  59.0  70.0  424.0   55.0   1.7  138.0  4.5  ...   
4       353025  3.0  4.0  40.0  70.0  253.0  150.0  11.9  132.0  5.6  ...   
..         ...  ...  ...   ...   ...    ...    ...   ...    ...  ...  ...   
148     486397  0.0  0.0  20.0  70.0  123.0   44.0   1.0  135.0  3.8  ...   
149     203466  0.0  0.0  22.0  60.0   97.0   18.0   1.2  138.0  4.3  ...   
150     482105  0.0  0.0  44.0  60.0   95.0   46.0   0.5  138.0  4.2  ...   
151     228849  0.0  0.0  58.0  80.0  100.0   50.0   1.2  140.0  3.5  ...   
152     996076  0.0  0.0  57.0  60.0  105.0   49.0   1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

It would be the same because there are only two values in binary - yes or no, true or false. If we have one column, we can infer the other because it has to be the other option. Like if yes = 1 then no must be 0. If there were to be more than 2 values then it would not be equivalent. If we have one column Mustard for example (out of 5 total columns), if it's value is 1 then we know that the other four have to be 0. However, if its value is 0 we have no way of knowing the values for the other four columns.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [55]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

def remove_outliers(data):
    for col in numerical_columns:
        mean = data[col].mean()
        std_dev = data[col].std()
        upperThreshold = mean + (3 * std_dev)
        lowerThreshold = mean - (3 * std_dev)
        data = data[(data[col] >= lowerThreshold) & (data[col] <= upperThreshold)]
    return data

filteredDataset = remove_outliers(encodedData)
filteredDataset.reset_index(drop=True, inplace=True)

# Print the dataset
print(filteredDataset)

     unique_id   al   su   age     bp    bgr     bu   sc    sod  pot  ...  \
0       343710  2.0  0.0  61.0   80.0  173.0  148.0  3.9  135.0  5.2  ...   
1       397388  2.0  2.0  63.0  100.0  280.0   35.0  3.2  143.0  3.5  ...   
2       438182  2.0  0.0  73.0   80.0  253.0  142.0  4.6  138.0  5.8  ...   
3       474407  4.0  0.0  21.0   90.0  107.0   40.0  1.7  125.0  3.5  ...   
4       475200  4.0  2.0  64.0  100.0  163.0   54.0  7.2  140.0  4.6  ...   
..         ...  ...  ...   ...    ...    ...    ...  ...    ...  ...  ...   
128     486397  0.0  0.0  20.0   70.0  123.0   44.0  1.0  135.0  3.8  ...   
129     203466  0.0  0.0  22.0   60.0   97.0   18.0  1.2  138.0  4.3  ...   
130     482105  0.0  0.0  44.0   60.0   95.0   46.0  0.5  138.0  4.2  ...   
131     228849  0.0  0.0  58.0   80.0  100.0   50.0  1.2  140.0  3.5  ...   
132     996076  0.0  0.0  57.0   60.0  105.0   49.0  1.2  150.0  4.7  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_ye

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [50]:
# Normalize the all Numerical Attributes in the dataset.

for col in numerical_columns:
    min = filteredDataset[col].min()
    max = filteredDataset[col].max()
    filteredDataset.loc[:, col] = (filteredDataset[col] - min) / (max - min)

# Print the dataset
print(filteredDataset)

     unique_id   al   su       age    bp       bgr        bu        sc  \
0       343710  2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099   
1       397388  2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679   
2       438182  2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519   
3       474407  4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494   
4       475200  4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506   
..         ...  ...  ...       ...   ...       ...       ...       ...   
128     486397  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074   
129     203466  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765   
130     482105  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346   
131     228849  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765   
132     996076  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765   

          sod       pot  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.500000  0.793103  ...   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The column for unique_id isn't appropriate for modeling and predictions because  it's unnecessary to know which specific id has this data, we want to model it to see the trend. 

In [51]:
#Remove that column
filteredDataset = filteredDataset.drop(columns="unique_id")

# Print the dataset
print(filteredDataset)

      al   su       age    bp       bgr        bu        sc       sod  \
0    2.0  0.0  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
1    2.0  2.0  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
2    2.0  0.0  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
3    4.0  0.0  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
4    4.0  2.0  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..   ...  ...       ...   ...       ...       ...       ...       ...   
128  0.0  0.0  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
129  0.0  0.0  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
130  0.0  0.0  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
131  0.0  0.0  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
132  0.0  0.0  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo  ...  cad_no  cad_yes  appet_good  appet_poor  pe_no  \
0    0.793103  0.000000  ...       0       

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [52]:
# Export the dataframe to a new csv file
csvFilename = 'cleaned_data.csv'
filteredDataset.to_csv(csvFilename, index=False)

# Export the dataframe to a new json file
jsonFilename = 'cleaned_data.json'
filteredDataset.to_json(jsonFilename, orient="records", indent=4)



# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Paste here: 

### Paste the markdown table here

| al   | su   | age      | bp   | bgr       | bu       | sc       | sod       | pot       | hemo      | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|------|------|----------|------|-----------|----------|----------|-----------|-----------|-----------|--------|---------|------------|-------------|-------|--------|--------|---------|------------|-----------------|
| 2.0  | 0.0  | 0.743243 | 0.50 | 0.442060  | 0.901961 | 0.432099 | 0.500000  | 0.657143  | 0.172131  | 0      | 1       | 0          | 1           | 0     | 1      | 0      | 1       | 1          | 0               |
| 2.0  | 2.0  | 0.770270 | 1.00 | 0.901288  | 0.163399 | 0.345679 | 0.766667  | 0.171429  | 0.606557  | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 2.0  | 0.0  | 0.905405 | 0.50 | 0.785408  | 0.862745 | 0.518519 | 0.600000  | 0.828571  | 0.401639  | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 4.0  | 0.0  | 0.202703 | 0.75 | 0.158798  | 0.196078 | 0.160494 | 0.166667  | 0.171429  | 0.221311  | 1      | 0       | 1          | 0           | 1     | 0      | 0      | 1       | 1          | 0               |
| 4.0  | 2.0  | 0.783784 | 1.00 | 0.399142  | 0.287582 | 0.839506 | 0.666667  | 0.485714  | 0.188525  | 1      | 0       | 1          | 0           | 0     | 1      | 1      | 0       | 1          | 0               |
| 3.0  | 1.0  | 0.729730 | 0.00 | 0.935622  | 0.169935 | 0.160494 | 0.333333  | 0.028571  | 0.188525  | 1      | 0       | 0          | 1           | 1     | 0      | 0      | 1       | 1          | 0               |
| 2.0  | 0.0  | 0.675676 | 0.75 | 0.253219  | 0.633987 | 0.777778 | 0.366667  | 0.542857  | 0.286885  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 4.0  | 3.0  | 0.851351 | 0.25 | 0.618026  | 0.562092 | 0.728395 | 0.000000  | 0.285714  | 0.311475  | 0      | 1       | 1          | 0           | 0     | 1      | 0      | 1       | 1          | 0               |
| 1.0  | 0.0  | 0.540541 | 0.00 | 0.399142  | 0.535948 | 0.358025 | 0.700000  | 0.314286  | 0.344262  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 3.0  | 0.0  | 0.756757 | 0.25 | 0.223176  | 0.209150 | 0.160494 | 0.533333  | 0.514286  | 0.573770  | 1      | 0       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 3.0  | 1.0  | 0.662162 | 0.50 | 0.618026  | 0.411765 | 0.432099 | 0.566667  | 0.571429  | 0.434426  | 1      | 0       | 1          | 0           | 0     | 1      | 1      | 0       | 1          | 0               |
| 4.0  | 1.0  | 0.783784 | 0.00 | 0.725322  | 0.313725 | 0.481481 | 0.566667  | 0.714286  | 0.319672  | 1      | 0       | 0          | 1           | 0     | 1      | 1      | 0       | 1          | 0               |
| 4.0  | 0.0  | 0.878378 | 0.00 | 0.206009  | 0.751634 | 0.604938 | 0.533333  | 0.571429  | 0.475410  | 1      | 0       | 0          | 1           | 0     | 1      | 1      | 0       | 1          | 0               |
| 3.0  | 0.0  | 0.878378 | 0.25 | 0.639485  | 0.470588 | 0.395062 | 0.433333  | 0.428571  | 0.393443  | 0      | 1       | 1          | 0           | 1     | 0      | 1      | 0       | 1          | 0               |
| 1.0  | 0.0  | 0.716216 | 0.50 | 1.000000  | 0.163399 | 0.111111 | 0.066667  | 0.171429  | 0.393443  | 1      | 0       | 0          | 1           | 1     | 0      | 1      | 0       | 1          | 0               |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

After conversion, we can cross check it with the original data to check for discrepancies. If dealing with a large dataset, we can break it into smaller chunks, and convert and validate chunk before moving on to the next. Use clear and detailed prompts when instructing the model for conversion tasks. We can use an iterative approach to progressively convert and validate parts of data, and adjust the conversion instructions before moving on. Also, we can document the process by keeping a detailed log of the conversion process like changes made and why to understand the progression. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [53]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns
filteredDataset.rename(columns=rename_mapping, inplace=True)
print(filteredDataset)

     albumin  sugar       age  blood pressure  blood glucose random  \
0        2.0    0.0  0.743243            0.50              0.442060   
1        2.0    2.0  0.770270            1.00              0.901288   
2        2.0    0.0  0.905405            0.50              0.785408   
3        4.0    0.0  0.202703            0.75              0.158798   
4        4.0    2.0  0.783784            1.00              0.399142   
..       ...    ...       ...             ...                   ...   
128      0.0    0.0  0.189189            0.25              0.227468   
129      0.0    0.0  0.216216            0.00              0.115880   
130      0.0    0.0  0.513514            0.00              0.107296   
131      0.0    0.0  0.702703            0.50              0.128755   
132      0.0    0.0  0.689189            0.00              0.150215   

     blood urea  serum creatinine    sodium  potassium  hemoglobin  ...  \
0      0.901961          0.432099  0.500000   0.793103    0.000000  ... 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [54]:
# Converted SQL to Pandas code

df_grouped = filteredDataset.groupby('Target_ckd').size().reset_index(name='count')
print(df_grouped)



   Target_ckd  count
0           0    111
1           1     22


*Describe what the above code does here*

The above code groups the data frame by the values in the Target_ckd column and it counts the occurrences for each unqiue values in that column. It returns a new Dataframe with two columns: the unique values from Target_ckd and the counts that correspond to each of them.